cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1828,6 +1828,148 @@ void AddUtf8Length(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
 
+// binary join
+
+template <typename Type>
+struct Join {
+  using ArrayType = typename TypeTraits<Type>::ArrayType;
+  using ListArrayType = ListArray;
+  using offset_type = typename Type::offset_type;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].kind() == Datum::SCALAR) {
+      const ListScalar& list = checked_cast<const ListScalar&>(*batch[0].scalar());
+      if (!list.is_valid) {
+        return;
+      }
+      if (batch[1].kind() == Datum::SCALAR) {
+        const BaseBinaryScalar& separator_scalar =
+            checked_cast<const BaseBinaryScalar&>(*batch[1].scalar());
+        if (!separator_scalar.is_valid) {
+          return;
+        }
+        util::string_view separator(*separator_scalar.value);
+
+        TypedBufferBuilder<uint8_t> builder(ctx->memory_pool());
+        auto Append = [&](util::string_view value) {
+          return builder.Append(reinterpret_cast<const uint8_t*>(value.data()),
+                                static_cast<offset_type>(value.size()));
+        };
+
+        const ArrayType* strings = static_cast<const ArrayType*>(list.value.get());
+        if (strings->null_count() > 0) {
+          // since the input list is not null, the out datum needs to be assigned to
+          *out = MakeNullScalar(list.value->type());
+          return;
+        }
+        if (strings->length() > 0) {
+          KERNEL_RETURN_IF_ERROR(ctx, Append(strings->GetView(0)));
+          for (int64_t j = 1; j < strings->length(); j++) {
+            KERNEL_RETURN_IF_ERROR(ctx, Append(separator));
+            KERNEL_RETURN_IF_ERROR(ctx, Append(strings->GetView(j)));
+          }
+        }
+        std::shared_ptr<Buffer> string_buffer;
+
+        KERNEL_RETURN_IF_ERROR(ctx, builder.Finish(&string_buffer));
+        KERNEL_ASSIGN_OR_RAISE(auto scalar_right_type, ctx,
+                               MakeScalar<std::shared_ptr<Buffer>>(
+                                   list.value->type(), std::move(string_buffer)));
+        *out = scalar_right_type;
+      }  // do we want to support scalar[list[str]] with array[str] ?
+    } else {
+      const ListArrayType list(batch[0].array());
+      ArrayData* output = out->mutable_array();
+
+      BuilderType builder(ctx->memory_pool());
+      KERNEL_RETURN_IF_ERROR(ctx, builder.Reserve(list.length()));
+      if (batch[1].kind() == Datum::ARRAY) {
+        ArrayType separator_array(batch[1].array());
+        for (int64_t i = 0; i < list.length(); ++i) {
+          const std::shared_ptr<Array> slice = list.value_slice(i);
```

Comment:
I think it would be much faster to avoid instantiating this temporary array for each list element.